saphir/src/server.rs

```diff
@@ -412,31 +391,65 @@ unsafe impl Send for Stack {}
 unsafe impl Sync for Stack {}
 
 impl Stack {
-    fn new_handler(&'static self, peer_addr: Option<SocketAddr>) -> StackHandler {
-        StackHandler { stack: self, peer_addr }
+    fn new_handler(&'static self, timeout_ms: Option<u64>, peer_addr: Option<SocketAddr>) -> StackHandler {
+        StackHandler {
+            timeout_ms,
+            stack: self,
+            peer_addr,
+        }
     }
 
-    async fn invoke(&self, mut req: Request<Body>) -> Result<Response<Body>, SaphirError> {
+    async fn invoke(&self, mut req: Request<Body>, timeout_ms: Option<u64>) -> Result<Response<Body>, SaphirError> {
+        use tokio::time::{timeout, Duration, Elapsed};
+
         let meta = self.router.resolve_metadata(&mut req);
+        let mut err_req = Request::clone_for_err(&req);
         let ctx = HttpContext::new(req, self.router.clone(), meta);
 
-        self.middlewares
-            .next(ctx)
+        if let Some(timeout_ms) = timeout_ms {
+            match timeout(Duration::from_millis(timeout_ms), async move {
+                self.middlewares
+                    .next(ctx)
+                    .await
+                    .and_then(|mut ctx| ctx.state.take_response().ok_or_else(|| SaphirError::ResponseMoved))
+            })
             .await
-            .and_then(|mut ctx| ctx.state.take_response().ok_or_else(|| SaphirError::ResponseMoved))
+            {
+                Ok(res) => res,
+                Err(Elapsed { .. }) => Err(SaphirError::RequestTimeout),
+            }
+        } else {
+            self.middlewares
+                .next(ctx)
+                .await
+                .and_then(|mut ctx| ctx.state.take_response().ok_or_else(|| SaphirError::ResponseMoved))
+        }
+        .or_else(|e| {
```

Comment:
The solution is as simple as adding an `or_else` combinator after the future to handle the `Err(Elapsed)` case. I don't understand why so much changes were needed, why 5 files got changed for this little solution. Furthermore you fix implies less performance by checking the timeout every time a request is spawned instead of before starting the server.

Reply:
The timeout need to be created on a per-request basis, not at server start. The fix is having a timeout on the invoke inside the Service impl passed to Hyper instead of wrapping the service itself in a timeout, because a timeout simply terminate the future without polling it to completion, which prevent returning a proper http response.